echo-pipelinetriggers/src/main/java/com/netflix/spinnaker/echo/pipelinetriggers/PipelineCache.java

```diff
@@ -63,6 +63,9 @@
   @Nullable
   private transient List<Pipeline> pipelines;
 
+  @Nullable
+  private transient Map<String, List<Trigger>> triggers;
```

Comment:
Should this (and possibly `pipelines` above) be `volatile` instead of (or in addition to) `transient`?  I think that would remove the race condition you note below about `triggers` still being null after calling `getPipelinesSync()` because the update by the other thread would be guaranteed to be visible.

Reply:
1. Why are these things marked `transient`, I don't think this class is ever serialized?
2. I don't know that `volatile` will fix the race - but it's probably a good thing to add anyway. I think it needs to either be a proper lock or we should change the order of how we assign pipeline/trigger. I.e. at https://github.com/spinnaker/echo/pull/526/files?file-filters%5B%5D=.java#diff-123d58ccbba1de00453419d04e10722eR139
```
tempPipelines = fetchHydratedPipelines();
triggers = extractEnabledTriggersFrom(tempPipelines);
pipelines = tempPipelined
```

not sure that I love either of these options

Reply:
also, maybe rename to `allTriggersByType`?

Reply:
Oops, I definitely meant `volatile` and not `transient`, good find 😅

After a good long read about the new guarantees in JSR-133... I sort of go back and forth between convincing myself that `volatile` fixes the race condition or not. In any case, I think it's safe to leave the `null` check in place and recompute `triggers` on the fly if that happens, that should be cheap enough.

Reply:
How about storing this in an AtomicReference instead? IMO that is much clearer and easier to get right than the equivalent level of `volatile` and `synchronized`

Reply:
I don't think `AtomicReference` buys us anything here since we're not doing operations like `compareAndSet`. From the docs (https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/package-summary.html):

> The memory effects for accesses and updates of atomics generally follow the rules for volatiles, as stated in section 17.4 of The Java™ Language Specification.
> 
> get has the memory effects of reading a volatile variable.
> set has the memory effects of writing (assigning) a volatile variable.